### PR TITLE
Improve S2259: ThrowHelper and Debug.Fail

### LINKED ISSUE #6081: Improve S2259: ThrowHelper and Debug.Fail
Some methods can be recognized as "always throwing". Such throw helpers are prominent in performance-critical scenarios to allow the JIT to inline methods (see e.g. https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-6/ "throw helper").

The following attributes can be used to identify such throw helpers
* [DoesNotReturnAttribute](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis#stop-nullable-analysis-when-called-method-throws)
* [TerminatesProgramAttribute﻿](https://www.jetbrains.com/help/resharper/Reference__Code_Annotation_Attributes.html#TerminatesProgramAttribute)

The following methods might not be annotated (in general or in older frameworks)but should be supported too:
* [Environment.FailFast](https://docs.microsoft.com/dotnet/api/system.environment.failfast)
* [Environment.Exit](https://docs.microsoft.com/dotnet/api/system.environment.exit)
* [Debug.Fail](https://docs.microsoft.com/dotnet/api/system.diagnostics.debug.fail)
* [ExceptionDispatchInfo.Throw](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.exceptionservices.exceptiondispatchinfo.throw?view=net-6.0#system-runtime-exceptionservices-exceptiondispatchinfo-throw)
* [NotFoundFileinfo.CreateReadStream](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.fileproviders.notfoundfileinfo.createreadstream?view=dotnet-plat-ext-6.0#microsoft-extensions-fileproviders-notfoundfileinfo-createreadstream)

see also #5824 for conditional annotations

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Fixes #6081

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs [hunk 1]
```diff
@@ -41,6 +42,16 @@ public static ProgramState Process(SymbolicContext context, IInvocationOperation
...
             }
+            state = ProcessThrowHelper(state, invocation);
+            return state;
+        }
+
+        private static ProgramState ProcessThrowHelper(ProgramState state, IInvocationOperationWrapper invocation)
+        {
+            if (KnownMethods.IsDebugFail(invocation.TargetMethod))
+            {
+                return ProgramState.Empty;
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs:53]: @pavel-mikula-sonarsource  The idea is to suppress any further processing. One option is to reset the state, but this does not stop the processing. A *check* is better suited, right? Except for `Environment.FailFast` (which skips catch, finally, and finalizers), all such methods raise exceptions. Can we simulate that from within a *check*?

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs:53]: Yes, InvocationCheck is a better place for now. We will eventually merge those two, but for now, the check it is. There should be something like `return EmptyStates` available. 
See #6118, you might want to base on that as it introduces ability to stop processing.

`Environment.FailFast` is pretty niche. We can just simply stop processing too, as there's nothing we will do about it (and nobody cares about the state of the program afterwards)

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Checks/InvocationCheck.cs [hunk 2]
```diff
@@ -26,27 +29,49 @@ namespace SonarAnalyzer.SymbolicExecution.Roslyn.Checks
 {
     internal class InvocationCheck : SymbolicCheck
     {
+        private static ImmutableArray<KnownType> throwHelperAttributes = new[]
+        {
+            KnownType.System_Diagnostics_CodeAnalysis_DoesNotReturnAttribute,
+            KnownType.JetBrains_Annotations_TerminatesProgramAttribute,
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Checks/InvocationCheck.cs:35]: There is an inconsistency here:
For "ValidatedNotNull" we only check for the class name but not the namespace:

`attribute.AttributeClass?.Name == "ValidatedNotNullAttribute"`

Here we also test for the namespace. What do you think is the better way forward for attribute support:
* [ ] always by name
* [ ] always by fully qualified name
* [ ] case-by-case

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 3]
```diff
@@ -413,5 +413,123 @@ public void Invocation_IsNullOrEmpty_TryFinally()
...
+    }}
+
+    static void Tag(string name) {{ }}
+}}
+
+namespace JetBrains.Annotations
+{{
+    [Obsolete(""Use [ContractAnnotation('=> halt')] instead"")]
+    [AttributeUsage(AttributeTargets.Method)]
+    public sealed class TerminatesProgramAttribute : Attribute {{ }}
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:508]: copied from [fuget](https://www.fuget.org/packages/JetBrains.Annotations/2022.1.0/lib/netstandard2.0/JetBrains.Annotations.dll/JetBrains.Annotations/TerminatesProgramAttribute?code=true)

`[ContractAnnotation]` support is also fairly easy to do here and we should consider it for the future.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Checks/InvocationCheck.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Checks/InvocationCheck.cs:35]: Here the attributes are framework specific and not the same (by coincidence). So well-known full name is the best way

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownMethods.cs [hunk 4]
```diff
@@ -158,6 +158,11 @@ public static bool IsGetObjectData(this IMethodSymbol methodSymbol)
             && methodSymbol.Name == nameof(Debug.Assert)
             && methodSymbol.ContainingType.Is(KnownType.System_Diagnostics_Debug);
 
+        public static bool IsDebugFail(this IMethodSymbol methodSymbol) =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownMethods.cs:161]: This is used just once. And there's `.Is(KnownType, Name)` extension available that we should use.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Checks/InvocationCheck.cs [hunk 5]
```diff
@@ -26,27 +29,49 @@ namespace SonarAnalyzer.SymbolicExecution.Roslyn.Checks
 {
     internal class InvocationCheck : SymbolicCheck
     {
+        private static ImmutableArray<KnownType> throwHelperAttributes = new[]
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Checks/InvocationCheck.cs:32]:
```suggestion
        private static readonly ImmutableArray<KnownType> ThrowHelperAttributes = new[]
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Checks/InvocationCheck.cs [hunk 6]
```diff
@@ -26,27 +29,49 @@ namespace SonarAnalyzer.SymbolicExecution.Roslyn.Checks
...
+        {
+            if (invocation.TargetMethod.IsAny(KnownType.System_String, nameof(string.IsNullOrEmpty), nameof(string.IsNullOrWhiteSpace)))
+            {
+                return ProcessStringIsNullOrEmpty(context.State, invocation);
+            }
+            else if (IsThrowHelper(invocation))
+            {
+                return EmptyStates; // Stop processing
+            }
+            return null;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Checks/InvocationCheck.cs:54]: Enforces every other branch to take decision and return
```suggestion
            else
            {
                return null;
            }               
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Checks/InvocationCheck.cs [hunk 7]
```diff
@@ -26,27 +29,49 @@ namespace SonarAnalyzer.SymbolicExecution.Roslyn.Checks
...
+        }
+
+        private static bool IsThrowHelper(IInvocationOperationWrapper invocation) =>
+            KnownMethods.IsDebugFail(invocation.TargetMethod)
+            || invocation.TargetMethod.IsAny(KnownType.System_Environment, nameof(Environment.FailFast), nameof(Environment.Exit))
+            || invocation.TargetMethod.GetAttributes(throwHelperAttributes).Any();
 
-        private static ProgramState[] ProcessStringIsNullOrEmpty(SymbolicContext context, IInvocationOperationWrapper invocation) =>
-            context.State[invocation.Arguments[0].ToArgument().Value]?.Constraint<ObjectConstraint>() switch
+        private static ProgramState[] ProcessStringIsNullOrEmpty(ProgramState state, IInvocationOperationWrapper invocation) =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Checks/InvocationCheck.cs:62]: Why this change? `context` was handy and this made the implementation header to read. Should be reverted

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 8]
```diff
@@ -413,5 +413,123 @@ public void Invocation_IsNullOrEmpty_TryFinally()
...
+
+        private static IEnumerable<object[]> ThrowHelperCalls
+        {
+            get
+            {
+                yield return new object[] { @"System.Diagnostics.Debug.Fail(""Fail"");" };
+                yield return new object[] { @"Environment.FailFast(""Fail"");" };
+                yield return new object[] { @"Environment.Exit(-1);" };
+            }
+        }
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:533]: Isn't it easier to do just arrow property instead, returning the array directly in this case?

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 9]
```diff
@@ -413,5 +413,123 @@ public void Invocation_IsNullOrEmpty_TryFinally()
...
+        // No implementation. The attribute should drive the analysis.
+    }}
+
+    static void Tag(string name) {{ }}
+}}
+
+namespace JetBrains.Annotations
+{{
+    [Obsolete(""Use [ContractAnnotation('=> halt')] instead"")]
+    [AttributeUsage(AttributeTargets.Method)]
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:507]: Should not be needed for our purposes (scaffolding noise)
```suggestion
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Minor polishing

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Checks/InvocationCheck.cs (hunk 7 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Checks/InvocationCheck.cs:62]: I don't like the idea to pass `SymbolicContext` as soon as there is more than one thing going on. You want to be able to do `state = DoSomeMore(state)` at one point and reorder it. It is hard to do if you pass `SymbolicContext`.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Checks/InvocationCheck.cs:62]: Content of this method is pretty much done in long run. The change made it harder to read.

File: analyzers/src/SonarAnalyzer.Common/Extensions/AttributeDataExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+using System;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+
+namespace SonarAnalyzer.Extensions
+{
+    public static class AttributeDataExtensions
+    {
+        public static bool HasName(this AttributeData attribute, string name)
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/AttributeDataExtensions.cs:29]: I added the extension method to `AttributeData` because we need this class to extract the arguments from attributes like `[NotNullWhen(false)]`. `AttributeData` contains the info, but it is a bit complicated to get it, so we need some helpers.

**Comment by martin-strecker-sonarsource** [comment]
I changed the name handling according to:
https://github.com/SonarSource/sonar-dotnet/pull/6110#discussion_r977770069

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 11]
```diff
@@ -413,5 +413,171 @@ public void Invocation_IsNullOrEmpty_TryFinally()
...
+";
+            var validator = SETestContext.CreateCS(code, ", bool condition").Validator;
+            validator.ValidateTagOrder("Catch", "Finally", "Finally", "End");
+            validator.ValidateExitReachCount(2);
+            validator.ValidateExecutionCompleted();
+        }
+
+        [DataTestMethod]
+        [DataRow("System.Diagnostics.CodeAnalysis.DoesNotReturn")]
+        [DataRow("OtherNamespace.WithAttributeSuffix.DoesNotReturn")]
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:479]: Based on the namespace name, shouldn't the name be like this?
```suggestion
        [DataRow("OtherNamespace.WithAttributeSuffix.DoesNotReturnAttribute")]
```

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 12]
```diff
@@ -413,5 +413,171 @@ public void Invocation_IsNullOrEmpty_TryFinally()
...
+        [DataRow("OtherNamespace.WithoutAttributeSuffix.DoesNotReturn")]
+        [DataRow("JetBrains.Annotations.TerminatesProgram")]
+        [DataRow("OtherNamespace.WithAttributeSuffix.TerminatesProgram")]
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:482]: Same here
```suggestion
        [DataRow("OtherNamespace.WithAttributeSuffix.TerminatesProgramAttribute")]
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Round 2, minor stuff

File: analyzers/src/SonarAnalyzer.Common/Extensions/AttributeDataExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+                throw new ArgumentException($"'{nameof(name)}' cannot be null or whitespace.", nameof(name));
+            }
+            if (name.EndsWith("attribute", comparision))
+            {
+                throw new ArgumentException($"The class name '{name}' should not have the Attribute suffix.", nameof(name));
+            }
+
+            return attribute is { AttributeClass.Name: { } attributeClassName }
+                && attributeClassName.StartsWith(name, comparision)
+                && (attributeClassName.Equals($"{name}Attribute", comparision) || attributeClassName.Equals(name, comparision));
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Extensions/AttributeDataExtensions.cs]: It is the other way around because we check the class name (Declaration). These usually have the Attribute suffix.

File: analyzers/src/SonarAnalyzer.Common/Extensions/AttributeDataExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        public static bool HasAnyName(this AttributeData attribute, params string[] otherNames) =>
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Extensions/AttributeDataExtensions.cs]: Leftover from some overloads I added to avoid the `string[]` instantiation for simple cases.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs [hunk 15]
```diff
@@ -61,4 +69,12 @@ protected override ProgramState[] Process(SymbolicContext context, IInvocationOp
             },
             _ => new[] { context.State }
         };
+
+    private static bool IsThrowHelper(IInvocationOperationWrapper invocation) =>
+        invocation.TargetMethod is { } method
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs:74]: > Can we craft an UT with // Error that would have CFG with Invocation without a TargetMethod?

I added an `f()()` invocation which is the closed I could come up with, but this has `Delegate.Invoke` as target. Also, the [nullable annotation in Roslyn](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.operations.iinvocationoperation?view=roslyn-dotnet-4.3.0) indicates that it is never null.

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/AttributeDataExtensionsTest.cs [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+using SonarAnalyzer.Extensions;
+
+namespace SonarAnalyzer.UnitTest.Extensions
+{
+    [TestClass]
+    public class AttributeDataExtensionsTest
+    {
+        [DataTestMethod]
+        [DataRow(true, "Test", "Test")]
+        [DataRow(true, "Test", "test")]
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Extensions/AttributeDataExtensionsTest.cs:32]: Yes. The idea is to support user-defined custom attributes that don't follow the naming convention (e.g. `[DoesnotReturn]`)

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs (hunk 11 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:479]: In this case, both names are valid.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs (hunk 12 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:482]: I added both variations where applicable.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs]: _For archeology:_ I tried `Undefined();` and it generates `InvalidOperation`

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs (hunk 11 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:480]: Yes, both are valid. But the UTs seems wrong to me. Originally, we had this pair
```
        [DataRow("OtherNamespace.WithAttributeSuffix.DoesNotReturn")]
        [DataRow("OtherNamespace.WithoutAttributeSuffix.DoesNotReturn")]
```
And I'm looking at the `WithAttributeSuffix` vs. `WithoutAttributeSuffix` declared intention (I like these namespaces). 

The point is, that `Without` is tested here
```
        [DataRow("OtherNamespace.WithoutAttributeSuffix.DoesNotReturn")]
                                                                     ^ No suffix here - all good
```
but the `With` didn't have it
```
        [DataRow("OtherNamespace.WithAttributeSuffix.DoesNotReturn")]
                                                                  ^ No suffix here either. And it should be IMO
```
So currently, we have these 3
```
        [DataRow("OtherNamespace.WithAttributeSuffix.DoesNotReturn")]   // <--- this one doesn't make sense, it's tested on the 3rd line
        [DataRow("OtherNamespace.WithAttributeSuffix.DoesNotReturnAttribute")]    // OK
        [DataRow("OtherNamespace.WithoutAttributeSuffix.DoesNotReturn")]          // OK
```

Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:480]: ...aaand now I see it, it's with suffix or without suffix on the class declaration.

While doable, I don't see a need to support `class DoesNotReturn`. Attributes should have `Attribute` suffix. And those we really need do have it. Let's remove it to simplify the logic.

File: analyzers/src/SonarAnalyzer.Common/Extensions/AttributeDataExtensions.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Extensions/AttributeDataExtensions.cs]: So it is to support `public class NotNull` without attribute suffix? I don't think we need that, let's remove it (same as below)

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/AttributeDataExtensionsTest.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Extensions/AttributeDataExtensionsTest.cs:32]: I don't see a need for that. Let's make it case-sensitive, also for VB.
While VB is case insensitive in terms of usage, the class names should be named propertly.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
No new comments, just answered unresolved threads above

File: analyzers/src/SonarAnalyzer.Common/Extensions/AttributeDataExtensions.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Extensions/AttributeDataExtensions.cs]: Fine with me. That is what the spec has to say about it:

> By convention, attribute classes are named with a suffix of Attribute.
[Source](https://learn.microsoft.com/dotnet/csharp/language-reference/language-specification/attributes#2121-general)

It is just a convention, and the compiler tolerates violations.

File: analyzers/src/SonarAnalyzer.Common/Extensions/AttributeDataExtensions.cs [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+using System;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+
+namespace SonarAnalyzer.Extensions
+{
+    public static class AttributeDataExtensions
+    {
+        public static bool HasName(this AttributeData attribute, string name) =>
+            attribute is { AttributeClass.Name: { } attributeClassName } && attributeClassName.Equals(name, StringComparison.Ordinal);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/AttributeDataExtensions.cs:30]: More streigh-forward
```suggestion
            attribute is { AttributeClass.Name: { } attributeClassName } && attributeClassName == name;
```

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

File: analyzers/src/SonarAnalyzer.Common/Extensions/AttributeDataExtensions.cs [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/AttributeDataExtensions.cs:21]:
```suggestion
```
